roll/ast.py

```diff
@@ -0,0 +1,684 @@
+        out.__next_id = self.__next_id
+        out.__next_scope_id = self.__next_scope_id
+        out.trace = self.trace
+        return out
+
+    @property
+    def next_id(self):
+        return self.__next_id
+
+    def pop_id(self):
+        id = self.__next_id
+        self.__next_id += 1
+        return id
+
+    @property
+    def next_scope_id(self):
+        return self.__next_scope_id
+
+    def pop_scope_id(self):
+        id = self.__next_scope_id
+        self.__next_scope_id -= 1
+        return id
+
+
+def trace(func):
+    """A decorator that updates the environment/counter to know its position in the current expression"""
+
+    def wrapper(*args, **kwargs):
+        # args[0] will be self (the expression token object)
+        # args[1] should always be the env/map object
+        args[1].trace.append(args[0])
+        out = func(*args, **kwargs)
+        args[1].trace.pop()
+        return out
+
+    return wrapper
+
+
+class Token:
+    """Returns a fully reduced version of the token"""
+
+    @trace
+    def reduce(self, env, counter):
+        raise NotImplementedError
+
+    """Returns a deep copy of the token where each variable with an ID in the map is replaced by a copy with the new ID"""
+
+    def substitute(self, old_to_new):
+        raise NotImplementedError
+
+    """Returns the type of the token, or raises a TypeError"""
+
+    def type(self, env):
+        raise NotImplementedError
+
+    """Recursively constructs a string representation of the token"""
+
+    def __str__(self):
+        raise NotImplementedError
+
+    """Attempts to deference the token if it is a pointer"""
+
+    def dereference(self, env):
+        return self
+
+    """Recursively sets the IDs of variables to be unique"""
+
+    @trace
+    def hash_vars(self, counter, map):
+        raise NotImplementedError
+
+    """Returns the value of the reduced token in Python primitives"""
+
+    @property
+    def pure(self):
+        raise NotImplementedError
+
+
+class TokenNumber(Token):
+    def __init__(self, value):
+        self.__value = value
+        logging.debug(self.__class__.__name__, self)
+
+    @trace
+    def reduce(self, env, counter):
+        return self
+
+    def substitute(self, _):
+        return TokenNumber(self.__value)
+
+    def type(self, _):
+        return TypeNumber
+
+    @trace
+    def hash_vars(self, counter, map):
+        pass
+
+    def __str__(self):
+        return str(self.__value)
+
+    @property
+    def pure(self):
+        return self.__value
+
+    def __eq__(self, other):
+        return isinstance(other, TokenNumber) and self.__value == other.__value
+
+
+class TokenString(Token):
+    def __init__(self, value):
+        self.__value = value
+        logging.debug(self.__class__.__name__, self)
+
+    @trace
+    def reduce(self, env, counter):
+        return self
+
+    def substitute(self, _):
+        return TokenString(self.__value)
+
+    def type(self, _):
+        return TypeString

```

Comment:
Missing an import here, too

Reply:
Both are remnants of a type system I'm yet to realise. I've removed it for now.